llvm/lib/CAS/MappedFileRegionArena.cpp

```diff
@@ -187,29 +253,45 @@ Expected<MappedFileRegionBumpPtr> MappedFileRegionBumpPtr::create(
     Result.Region = std::move(Map);
   }
 
-  if (FileSize->Size == 0) {
-    assert(InitLock.Locked == sys::fs::LockKind::Exclusive);
-    // We are creating a new file; run the constructor.
+  // Initialize the header.
+  Result.initializeHeader(HeaderOffset);
+
+  if (FileSize->Size < MinCapacity) {
+    assert(MainFile->Locked == sys::fs::LockKind::Exclusive);
+    // If we need to fully initialize the file, call NewFileConstructor.
     if (Error E = NewFileConstructor(Result))
       return std::move(E);
-  } else {
-    Result.initializeBumpPtr(BumpPtrOffset);
+
+    Result.H->HeaderOffset.exchange(HeaderOffset);
+    Result.H->Capacity.exchange(Capacity);
+
+    // Sync to make sure other processes can read using read/fread from file
+    // system. Only sync the beginning of the file.
```

Comment:
@benlangmuir I am not sure if this `sync` is strictly needed but feels like it is safe to have.